### DISCUSSION TITLE
2022/09/01/beware-of-records-with-expressions-and-calculated-properties/

### DISCUSSION THREAD

**Discussion by giscus** [CodingMilitia/Blog#92] [Blog comments]
# 2022/09/01/beware-of-records-with-expressions-and-calculated-properties/

I’ve been using C# records a lot since the feature was introduced. However, when using them, we really need to understand how they work, otherwise we might face unexpected surprises.

https://blog.codingmilitia.com/2022/09/01/beware-of-records-with-expressions-and-calculated-properties/

**Comment by Rast1234** [comment]
Nice wrap up, too bad i found this the hard way, after debugging for several hours. Another problem is that we break equality with any of these approaches: automatically implemented Equals and GetHashCode use all fields and properties, even private ones. So, while we can have a private field for Lazy or for a raw cached value which is initialized once, all that it changes Equals and GetHashCode, breaking record value-semantics and read-onliness.


The obvious way around that is to implement Equals and GethashCode yourself, excluding private fields used for caching computed values, which is even worse...
Another way would be to implement setters for all the properties which relate to computed one, so it is eagerly calculated with every `with { }` call, but it's not good either.

After screwing around, i just went with a get-only auto computed property because perf was not as important as readability.

**Reply by joaofbantunes** [reply]
Great point! I completely forgot about the equality, as I wasn't using it in the particular case I encountered this issue.
